OA-PASS/pass-docker
# Add policy service and update proxies

# To Test
* `docker-compose up -d`
* Log into Ember `https://pass.local/app` as somebody (e.g. `nih-user`), look at the submissions table, and look at a submission.  Copy its URL, and paste into the browser `https://pass.local/policyservice/policies?submission=${submission}`, for example https://pass.local/policyservice/policies?submission=https://pass.local/fcrepo/rest/submissions/04/4c/0a/2e/044c0a2e-43fa-4611-889a-36fd7c104161
  * You should see one institutional policy and, if there any grants tied to funders with policies, funder policies as well.  The above example has one of each.
* Now try looking at the same page directly (not via shib), as `http://localhost:8088/policies?submission=${submission}`, for example http://localhost:8088/policies?submission=https://pass.local/fcrepo/rest/submissions/04/4c/0a/2e/044c0a2e-43fa-4611-889a-36fd7c104161 (or, use whatever docker-machine IP is appropriate).  You should see the institutional policy disappear, as that policy is determined by matching headers, and the shib headers are not present when hitting the service directly (see [the configuration](https://github.com/OA-PASS/pass-policy-service/blob/master/policies/docker.json#L28-L34) for why)
* Now hit the `/repositories` endpoint: https://pass.local/policyservice/repositories?submission=https://pass.local/fcrepo/rest/submissions/04/4c/0a/2e/044c0a2e-43fa-4611-889a-36fd7c104161
  * Oops, why is everything empty?  It's because we forgot to set `effectivePolicies` on the submission. 
* We need to set the value of effectivePolicies.  Go to the [fedora resource](https://pass.local/fcrepo/rest/submissions/04/4c/0a/2e/044c0a2e-43fa-4611-889a-36fd7c104161) in the browser.  In the "Update properties" box (you may need to press the red toggle button if it is not visible", and paste the following:
```
PREFIX pass: <http://oapass.org/ns/pass#>
PREFIX dcjsonld: <http://dataconservancy.org/ns/jsonld#>

DELETE { <> dcjsonld:hasContext ?o}
INSERT { 
<> pass:effectivePolicies <https://pass.local/fcrepo/rest/policies/e7/3f/26/70/e73f2670-6ef6-4201-bbcd-04631a93d852>.
<> pass:effectivePolicies <https://pass.local/fcrepo/rest/policies/5e/2e/16/92/5e2e1692-c128-4fb4-b1a0-95c0e355defd> .
<> dcjsonld:hasContext <https://oa-pass.github.io/pass-data-model/src/main/resources/context-3.4.jsonld>
}
WHERE { ?s ?p ?o }
```
* _now_ hit the repositories endpoint: https://pass.local/policyservice/repositories?submission=https://pass.local/fcrepo/rest/submissions/04/4c/0a/2e/044c0a2e-43fa-4611-889a-36fd7c104161
  * you should see one "required" repository, and one "optional" repository
* Hit the endpoint raw (not through shib): http://localhost:8088/repositories?submission=https://pass.local/fcrepo/rest/submissions/04/4c/0a/2e/044c0a2e-43fa-4611-889a-36fd7c104161
  * Why does it fail now?  Because of the lack of an eppn header.  The policy service calculates the list of applicable policies for a given resource, and expects `effectivePolicies` to be a subset of that.  In other words, you can't add any random policy via `effectivePolicies`.  In this case, since the request does not come from shibboleth, the policy service does not consider the j10p policy a valid policy for the submission, and spikes the submission which _does_ claim to satisfy it.

Resolves #198
